spec/services/reblog_service_spec.rb

```diff
@@ -37,11 +37,22 @@
   context 'when the reblogged status is discarded in the meantime' do
     let(:status) { Fabricate(:status, account: alice, visibility: :public) }
 
+    # Add a callback to discard the status being reblogged after the
+    # validations pass but before the database commit is executed.
     before do
-      # Update the in-database attribute without reflecting the change in
-      # the object. This cannot simulate all race conditions, but it is
-      # pretty close.
-      Status.where(id: status.id).update_all(deleted_at: Time.now.utc) # rubocop:disable Rails/SkipsModelValidations
+      Status.class_eval do
+        before_save :discard_status
+        def discard_status
+          Status
+            .where(id: reblog_of_id)
+            .update_all(deleted_at: Time.now.utc) # rubocop:disable Rails/SkipsModelValidations
```

Comment:
This should check that `reblog_of_id` is not `nil`, and/or have another way to identify the expected status (e.g. what you did before)